Tasks/Common/azure-arm-rest/azure-arm-mysql.ts

```diff
@@ -0,0 +1,320 @@
+        if (!callback) {
+            throw new Error(tl.loc("CallbackCannotBeNull"));
+        }
+
+        // Validate
+        try {
+            this.client.isValidResourceGroupName(resourceGroupName);
+            if(!this.isNameValid(serverName)){
+                throw new Error(tl.loc("MysqlServerNameCannotBeEmpty"));
+            }
+            if(!this.isNameValid(firewallRuleName)){
+                throw new Error(tl.loc("FirewallRuleNameCannotBeNull"));
+            }
+        }
+        catch (error) {
+            return callback(error);
+        }
+
+        var httpRequest = new webClient.WebRequest();
+        httpRequest.method = 'GET';
+        httpRequest.headers = {};
+        httpRequest.uri = this.client.getRequestUri(
+            '//subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforMySQL/servers/{serverName}/firewallRules/{firewallRuleName}',
+            {
+                '{resourceGroupName}': resourceGroupName,
+                '{serverName}': serverName,
+                '{firewallRuleName}': firewallRuleName
+            });
+        tl.debug("Calling get firewall ");
+        this.client.beginRequest(httpRequest).then(async (response) => {
+            var deferred = Q.defer<azureServiceClient.ApiResult>();
+            var statusCode = response.statusCode;
+            tl.debug("Response for get firewall rule " +statusCode);
+            if (statusCode === 200) {
+                // Generate Response
+                tl.debug("Response for get firewall rule " + JSON.stringify(response));
+                deferred.resolve(new azureServiceClient.ApiResult(null, response));
+            }
+            else {
+                // Generate exception
+                deferred.reject(new azureServiceClient.ApiResult(azureServiceClient.ToError(response)));
+            }
+            return deferred.promise;
+        }).then((apiResult: azureServiceClient.ApiResult) => callback(apiResult.error, apiResult.result),
+            (error) => callback(error));
+    }
+
+
+    /**
+     * Retry get call to check firewall rule has added or not
+     * @param resourceGroupName     resource group name of mysql server 
+     * @param serverName            mysql server name
+     * @param firewallRuleName      firewall rule name to be deleted 
+     * @param retryOption           no of time to retry
+     */
+    private _recursiveGetCall(resourceGroupName: string, serverName: string, firewallRuleName: string, retryOption: number) : Q.Promise<azureServiceClient.ApiResult>{
+        var deferred = Q.defer<azureServiceClient.ApiResult>();
+        setTimeout(() => {
+            this.get(resourceGroupName, serverName, firewallRuleName, (error, result, request, response) => {
+                if(error){
+                    if(retryOption > 0){
+                        deferred.resolve(this._recursiveGetCall(resourceGroupName, serverName, firewallRuleName, retryOption--));
```

Comment:
not sure about this line.
We are resolving a promise with a promise.
shouldn't we have like `defer.resolve(await this._recursiveCall)`?

Reply:
also not sure about  `retryOption--` , the value might not be updated in next iteration, leading to memory flow.
Can we provide something like `retryOption-1`?

Reply:
The flow will be like this first call for this method give you promise and when it will receive the response there will be two case if it is completed successfully it will resolve this one other wise it will recursively call with same flow .

Eg: call 1
case 1 success:- resolve this one
case 2 error :- already tried 3 times reject this promise
                       else resolve this promise with the response returned by recursive call of same method which is again promise.
In worst case when call3 will resolve then callback will reach to call2 resolve method and finally reach to call1 resolve method. It is simple recursive approach. It is working fine for all case.

If we will use  defer.resolve(await this._recursiveCall) then we are blocking our main thread which we completely like sync call.

Reply:
For retryOptions-- to retryOptions -1 I am taking it.